node/src/database/archive.rs

```diff
@@ -0,0 +1,316 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use std::fs;
+use std::path::Path;
+
+use anyhow::Result;
+use node_data::events::contract::ContractTxEvent;
+use node_data::ledger::Hash;
+use sqlx::sqlite::{SqliteConnectOptions, SqlitePool};
+use tracing::{info, warn};
+
+use super::Archivist;
+
+const ARCHIVE_FOLDER_NAME: &str = "archive";
+const SQLITE_DB_NAME: &str = "archive.sqlite3";
+
+#[derive(Debug)]
+pub struct SQLiteArchive {
+    archive_db: SqlitePool,
+}
+
+impl SQLiteArchive {
+    pub async fn create_or_open<T>(path: T) -> Self
+    where
+        T: AsRef<Path>,
+    {
+        let path = path.as_ref().join(ARCHIVE_FOLDER_NAME);
+        info!("Opening archive db in {path:?}");
+
+        // Recursively create the archive folder if it doesn't exist already
+        fs::create_dir_all(&path)
+            .expect("creating directory in {path} should not fail");
+
+        let db_options = SqliteConnectOptions::new()
+            // append the database name to the path
+            .filename(path.join(SQLITE_DB_NAME))
+            .create_if_missing(true);
+
+        // Open the database, create it if it doesn't exist
+        let archive_db = SqlitePool::connect_with(db_options)
+            .await
+            .expect("Failed to open database");
+
+        // Run the migrations
+        sqlx::migrate!("./migrations")
+            .run(&archive_db)
+            .await
+            .expect("Failed to run migrations");
+
+        Self { archive_db }
+    }
+
+    /// Fetch the json string of all vm events form a given block height
+    pub async fn fetch_json_vm_events(
+        &self,
+        block_height: u64,
+    ) -> Result<String> {
+        let block_height: i64 = block_height as i64;
+
+        let mut conn = self.archive_db.acquire().await?;
+
+        let events = sqlx::query!(
+            r#"SELECT json_contract_events FROM archive WHERE block_height = ?"#,
+            block_height
+        ).fetch_one(&mut *conn).await?;
+
+        Ok(events.json_contract_events)
+    }
+}
+
+impl Archivist for SQLiteArchive {
+    /// Store the list of all vm events from the block of the given height.
+    async fn store_vm_events(
+        &self,
+        block_height: u64,
+        block_hash: Hash,
+        events: Vec<ContractTxEvent>,
+    ) -> Result<()> {
+        let block_height: i64 = block_height as i64;
+        let hex_block_hash = hex::encode(block_hash);
+        // Serialize the events to a json string
+        let json_contract_events = serde_json::to_string(&events).unwrap();
```

Comment:
Maybe its okay to panic here since its related to the database? Ideally having panics in production is not good IMO, and if its something related to the node running on clusters, it can work as a DDOS attack

Reply:
good catch